Documentation/guides/internals/debug-jni-objrefs.md

```diff
@@ -0,0 +1,400 @@
+
+
+		// Invoke `Activity.runOnUiThread(Runnable)`, using the now unassociated object reference `h`
+		JniArgumentValue* args = stackalloc JniArgumentValue[1];
+		args[0] = new JniArgumentValue(h);
+		this.JniPeerMembers.InstanceMethods.InvokeVirtualVoidMethod("runOnUiThread.(Ljava/lang/Runnable;)V", this, args);
+	}
+}
+```
+
+
+<a name="crash-dalvik-abort"></a>
+
+### Crash via Dalvik Abort
+
+When debugging an app crashing via Dalvik Abort, the app immediately stops debugging.  There is no unhandled exception dialog; there is no exception.
+
+The Application Output window will contain `adb logcat` output similar to:
+
+```
+java_vm_ext.cc:579] JNI DETECTED ERROR IN APPLICATION: JNI ERROR (app bug): jobject is an invalid local reference: 0x75 (deleted reference at index 7 in a table of size 7)
+```
+
+This happens when an invalid JNI object reference is used.  This should *never happen* in normal use, but
+can be caused by deliberately misusing things:
+
+```csharp
+static unsafe void UseInvalidJniHandle()
+{
+	var a = new Java.Lang.String("a");
+	var b = JniEnvironment.Strings.NewString("b");
+	var h = b.Handle;
+	JniObjectReference.Dispose(ref b);
+	b = new JniObjectReference(h);
+	JniArgumentValue* args = stackalloc JniArgumentValue[1];
+	args[0] = new JniArgumentValue(b);
+	a.JniPeerMembers.InstanceMethods.InvokeNonvirtualBooleanMethod("endsWith.(Ljava/lang/String;)Z", a, args);
+}
+```
+
+The JNI Object reference `h` is invalidated when `JniObjectReference.Dispose(ref b)` is invoked; `h` is now invalid.
+Attempting to subsequently use that value results in the `JNI DETECTED ERROR IN APPLICATION` message and subsequent crash.
+
+
+<a name="collect-logs"></a>
+
+## Collect JNI Object Reference Logs
+
+In order to diagnose and fix JNI Object Reference usage bugs, JNI Object Reference logs must first be collected.
+There are two sets of logs: JNI *Local* Reference logs, and JNI *Global* Reference logs.  JNI Local Reference logs
+are *incredibly* voluminous; they should be collected only as a last resort.
+
+There are two ways to obtain JNI Object Reference logs:
+
+ 1. [Collect the *complete* JNI Object Reference log output](#collect-complete-logs), or
+ 2. [Collect "best effort" JNI Object Reference log output](#collect-best-effort-logs).
+
+
+<a name="collect-complete-logs"></a>
+
+### Collect Complete JNI Object Reference Logs
+
+To collect *complete* JNI object reference logs, your app must be "debuggable": the
+[`//application/@android:debuggable`](https://developer.android.com/guide/topics/manifest/application-element#debug) attribute
+within `AndroidManifest.xml` must have the value `true`.  This is typically the case for Debug configuration builds, and
+*not* for Release configuration builds.  (The Google Play Store requires that submitted apps *not* be debuggable.)
+
+Enable JNI Global Reference log collection by setting the `debug.mono.log` system property to a value which contains `gref`:
+
+	adb shell setprop debug.mono.log gref
```

Comment:
I'd suggest adding

```shell
adb logcat -G 16M
```

to increase the logcat buffer size, otherwise the logcat daemon will truncate messages